core/amber/src/main/scala/edu/uci/ics/texera/web/SyncableState.scala

```diff
@@ -0,0 +1,61 @@
+package edu.uci.ics.texera.web
+
+import rx.lang.scala.Subscription
+
+import java.util.concurrent.locks.ReentrantLock
+import scala.collection.mutable
+
+class SyncableState[T](defaultStateGen: => T) {
+
+  private var state: T = defaultStateGen
+  private var isModifying = false
+  private var callbackId = 0
+  private var onChangedCallbacks = mutable.HashMap[Int, (T, T) => Unit]()
+  private val lock = new ReentrantLock()
+
+  def getStateThenConsume[X](next: T => X): X = {
+    withLock {
+      next(state)
+    }
+  }
+
+  def acquireLock(): Unit = lock.lock()
+  def releaseLock(): Unit = lock.unlock()
+
+  private def withLock[X](code: => X): X = {
```

Comment:
this can be separated to some common/util package with implicit lock?

Reply:
the name `code` is not very friendly. maybe `statements` or `func` or `instructions` etc?

Reply:
moved it to util and change `code` to `instructions`